pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -584,3 +593,163 @@ func (a *Actuator) nodeAddresses(host *bmh.BareMetalHost) ([]corev1.NodeAddress,
+		_, exists = baremetalhost.Annotations[rebootAnnotation]
+	}
+	return
+}
+
+//addRemediationInProgressAnnotation adds a remediation-in-progress annotation to the machine
+func (a *Actuator) addRemediationInProgressAnnotation(ctx context.Context, machine *machinev1.Machine) error {
+	if machine.Annotations == nil {
+		machine.Annotations = make(map[string]string)
+	}
+
+	machine.Annotations[remediationInProgressAnnotation] = ""
+
+	err := a.client.Update(ctx, machine)
+	if err != nil {
+		log.Printf("Failed to add remediation in progess annotation to %s: %s", machine.Name, err.Error())
+	}
+
+	return err
+}
+
+//requestPowerOn removes reboot annotation on baremetalhost which signal BMO to power on the machine
+func (a *Actuator) requestPowerOff(ctx context.Context, baremetalhost *bmh.BareMetalHost) error {
+	if baremetalhost.Annotations == nil {
+		baremetalhost.Annotations = make(map[string]string)
+	}
+
+	baremetalhost.Annotations[rebootAnnotation] = ""
+
+	err := a.client.Update(ctx, baremetalhost)
+	if err != nil {
+		log.Printf("failed to add reboot annotation to %s: %s", baremetalhost.Name, err.Error())
+	}
+
+	return err
+}
+
+//requestPowerOn adds reboot annotation on baremetalhost which signal BMO to power off the machine
+func (a *Actuator) requestPowerOn(ctx context.Context, baremetalhost *bmh.BareMetalHost) error {
+	if baremetalhost.Annotations == nil {
+		baremetalhost.Annotations = make(map[string]string)
+	}
+
+	if _, rebootPending := baremetalhost.Annotations[rebootAnnotation]; !rebootPending {
+		return &clustererror.RequeueAfterError{}
+	}
+
+	delete(baremetalhost.Annotations, rebootAnnotation)
+
+	err := a.client.Update(ctx, baremetalhost)
+	if err != nil {
+		log.Printf("failed to remove reboot annotation from %s: %s", baremetalhost.Name, err.Error())
+	}
+
+	return err
+}
+
+// deleteMachineNode deletes the node that mapped to specified machine
+func (a *Actuator) deleteNode(ctx context.Context, node *corev1.Node) error {
+	err := a.client.Delete(ctx, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return &clustererror.RequeueAfterError{}
+		}
+		log.Printf("Failed to delete node %s: %s", node.Name, err.Error())
+	}
+	return err
+}
+
+// getNodeByMachine returns the node object referenced by machine
+func (a *Actuator) getNodeByMachine(ctx context.Context, machine *machinev1.Machine) (*corev1.Node, error) {
+	if machine.Status.NodeRef == nil {
+		return nil, errors.NewNotFound(corev1.Resource("ObjectReference"), machine.Name)
+	}
+
+	node := &corev1.Node{}
+	key := client.ObjectKey{
+		Name:      machine.Status.NodeRef.Name,
+		Namespace: machine.Status.NodeRef.Namespace,
+	}
+
+	if err := a.client.Get(ctx, key, node); err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+//remediateIfNeeded will try to remediate unhealthy nodes (annotated by MHC)
+func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1.Machine, baremetalhost *bmh.BareMetalHost) error {
```

Comment:
I find it hard to keep track of what's true or not in this function and where we are in the workflow. I made some naming suggestions regarding the annotations that I think will help future readers of this code (including future us!) understand what's going on.

Additionally, it would help a lot to write out in plain English, in the comment on this function, what the expected full workflow is.

Reply:
Yep, I agree. I'll add some comments to improve readability

Reply:
Added, let me know if you think that anything else is required